evalml/automl/automl_search.py

```diff
@@ -720,6 +724,7 @@ def _post_evaluation_callback(self, pipeline, evaluation_results, job_log):
             "percent_better_than_baseline": percent_better_than_baseline[self.objective.name],
             "validation_score": cv_scores[0]
         }
+        self._pipelines_searched.update({pipeline_id: pipeline})
```

Comment:
@angela97lin @chukarsten @freddyaboulton @bchen1116 : tagging you all here since I see there was a discussion earlier in this file.

Trained pipelines can grow quite large, particularly tree-based models ([one example here](https://stackoverflow.com/questions/61591998/do-xgboost-models-tend-to-be-large-in-size-mb-on-disk)). If we keep references to trained pipelines, our memory consumption will grow linearly over the course of automl search, and could grow to be quite large if enough pipelines are evaluated.

We shouldn't save trained pipelines here.

However, all we really need is the ability to construct an untrained copy of the pipeline. So I have two thoughts on how we can achieve that here:
1. Save an untrained copy of the pipeline here instead of a trained copy
```python
self._pipelines_searched.update({pipeline_id: pipeline.clone()})
```
The code in `get_pipeline` below should still work fine because it calls `PipelineBase.new` to get a new instance, so that's unaffected.

2. Avoid the need for the `self._pipelines_searched` dict entirely by saving an index into `allowed_pipelines` in the pipeline results, like so:
```python
self._results['pipeline_results'][pipeline_id] = {
   ...,
   "pipeline_instance_index": allowed_pipelines.index(pipeline)
}
```
Both of those approaches have pros and cons, but IMO they are both preferable to saving trained pipelines.

What do you think @angela97lin ? I think option 1 is fine, and is not far from what you already have in place.

Reply:
This is a really good call, thanks @dsherry! I agree that this is preferable to saving trained pipelines.

I opted for option 1, which is closer to what I was initially thinking of (storing a dictionary or something of the info a pipeline needs). I think option 2 is a little bit tricker--`allowed_pipelines` holds our original batch of pipelines, not all of our pipelines.

Reply:
Sounds good! Thanks :)

> allowed_pipelines holds our original batch of pipelines, not all of our pipelines.
Ah right. Its confusing... the `allowed_pipelines` hold the complete set of component graphs which automl will search. But yes as you said, those particular pipelines have an initial set of parameters, which happens to correspond to what our first batch will use. We will definitely have a chance to revisit this organization when we're ready to make some automl algo improvements.

Reply:
Yes, agreed! Perhaps when we move towards holding just graph representations, we don't need to store the pipelines at all, and just need to keep track of the parameters as we search (#2159)